Cheap ordering of elements with Hilbert spatial sort [hilbert-sort-dev]

jakubcerveny opened the pull request:
Adds the method `Mesh::GetHilbertElementOrdering`, which performs a spatial sort of mesh elements, to be used for coarse mesh reordering. This is a cheap (but fast) alternative to `GetGeckoElementOrdering`, when preparing the mesh for NC AMR. Could also be useful to improve caching. 

The elements are simply mapped to the Hilbert curve on the unit square. The result is not perfect but it is way better than no ordering, and for many applications it may be sufficient. 

Example ordering:
![hilbert-sort-0](https://user-images.githubusercontent.com/1349371/69267478-81acad00-0bcd-11ea-99a2-5edb248afca6.png)

When used for cheap partitioning (if we ever wanted to get rid of METIS as a mandatory dependency):

![hilbert-sort-1](https://user-images.githubusercontent.com/1349371/69267590-c8020c00-0bcd-11ea-87bf-75120054f49d.png)
![hilbert-sort-2](https://user-images.githubusercontent.com/1349371/69267597-cb959300-0bcd-11ea-8d5b-4b1eb92dbf46.png)

TODO:
* [x] Support for 3D meshes
* [x] Update CHANGELOG

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1166 | @jakubcerveny | @v-dobrev | @rw-anderson + @bslazarov | 12/04/19 | 12/19/19 | ⌛due 12/28/19 |

jakubcerveny commented:
Note: SFC-based partitioning tends to work better as the number of elements per process gets smaller:

![hilbert-sort-3](https://user-images.githubusercontent.com/1349371/69272897-a574f080-0bd7-11ea-8023-3d76c110aab8.png)

jakubcerveny commented:
Added support for 3D sorting.

![hilbert-sort-6](https://user-images.githubusercontent.com/1349371/69369871-17fbd400-0c9d-11ea-832e-66a172617492.png)
![hilbert-sort-4](https://user-images.githubusercontent.com/1349371/69369881-1d591e80-0c9d-11ea-8e46-8a7cbfefb1bb.png)

tzanio commented:
Oh, these are cool and may very well be sufficient for our purposes... @Dan2997925 can you give it a try in the problem where previously you had issues?

rw-anderson commented:
Nice idea.  How is the scale of the Hilbert curve chosen?

jakubcerveny commented:
The curve is mapped into the bounding box of the mesh. The question is whether this is OK or if the box should actually have equal sides. It seems to works OK on the pipe with a tight box...

rw-anderson commented:
I was wondering more about how many recursion levels on the Hilbert curve there are within the bounding box.  Or maybe I am misunderstanding something.

jakubcerveny commented:
There are two implementations actually, the first (2D only and disabled atm) converts a position into a distance along the curve (`PointToHilbertIndex`) and the elements are eventually sorted linearly by the index. This version uses 20 levels of the curve, or 20 bits per coordinate (60 in total for a potential 3D index). The second implementation does not construct the index explicitly and instead does something like a spatial QuickSort: starting at the top level, the points are partitioned into 8 octants by 3 midplanes. The function calls itself recursively for the octants (with some inversions and rotations of the coordinates), if they still contain some points. This version is only limited by the precision of the coordinates (53 bits?). I chose the second version even though the code is probably longer.

jakubcerveny commented:
I think this is ready for review. The demo code in `ex1` and `ex1p` will be removed after review.

tzanio commented on examples/ex1p.cpp:
```diff
@@ -122,19 +122,36 @@ int main(int argc, char *argv[])
+   Array<int> ordering;
+   mesh->GetHilbertElementOrdering(ordering);
+   mesh->ReorderElements(ordering);
+
```
I like this interface... should be very easy for our users

rw-anderson commented on mesh/mesh.cpp:
```diff
@@ -1383,6 +1383,177 @@ void Mesh::GetGeckoElementReordering(Array<int> &ordering,
+   }
+};
+
+static void HilbertSort2D(int coord1, bool dir1, bool dir2,
```
Nice algorithm here.  Looks like dir1 and dir2 relate to rotating and flipping the quadrants upon recursion, but their meaning isn't that obvious at the outset.  I don't know if there's a comment that would make it obvious or not.

jakubcerveny replied:
Added a couple comments related to this.

v-dobrev commented:
We should probably add a menu option in the `mesh-explorer` miniapp to apply this reordering. The code there can serve as illustration for users. The Gecko-based reordering can be added there too.

jakubcerveny commented:
@bslazarov Thanks for testing. I will add an ordering option in `mesh-explorer`.

The algorithm should scale very well. This is basically QuickSort, so something like O(n*log(n)). You can use this on millions of elements -- whatever fits in memory.

tzanio commented:
@bslazarov and @jakubcerveny -- should we wait for the ordering option in `mesh-explorer` or is this approved and ready for `next` as it is?

jakubcerveny commented:
Added a new submenu to reorder the mesh in `mesh-explorer`. Gecko will go there in the future too. Also added the option to partition the mesh along the 1D sequence of all elements.

This was done in `mesh-expolorer` using the above:
![explorer](https://user-images.githubusercontent.com/1349371/71099938-6d4fd600-21b4-11ea-8999-5062ee7863ba.png)

tzanio commented:
@jakubcerveny, can you mention the new `mesh-explorer` option in `CHANGELOG`?

@bslazarov, can you take another look and approve?

tzanio commented:
Merged in `next` for testing ...

tzanio commented:
Unit testing fails as shown below. Can you fix please.

See the [full log](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/19cd0772c751644e263ffebe335fc77c3276c084#tux426/2019-12-20-llnl-gh-next/unit-test.out)) in mfem/autotest.

```
Building MFEM with Gecko ...
Building the unit tests ...
rm -f *.o */*.o */*~ *~
g++ -g -c unit_test_main.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o unit_test_main.o
g++ -g -c fem/test_1d_bilininteg.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_1d_bilininteg.o
g++ -g -c fem/test_2d_bilininteg.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_2d_bilininteg.o
g++ -g -c fem/test_3d_bilininteg.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_3d_bilininteg.o
g++ -g -c fem/test_calcshape.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_calcshape.o
g++ -g -c fem/test_datacollection.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_datacollection.o
g++ -g -c fem/test_fe.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_fe.o
g++ -g -c fem/test_intrules.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_intrules.o
g++ -g -c fem/test_intruletypes.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_intruletypes.o
g++ -g -c fem/test_inversetransform.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_inversetransform.o
g++ -g -c fem/test_lin_interp.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_lin_interp.o
g++ -g -c fem/test_linear_fes.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_linear_fes.o
g++ -g -c fem/test_quadraturefunc.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o fem/test_quadraturefunc.o
g++ -g -c linalg/test_blockMatrix.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o linalg/test_blockMatrix.o
g++ -g -c linalg/test_densematrix.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o linalg/test_densematrix.o
g++ -g -c mesh/test_mesh.cpp -I.  -g -Wall -std=c++11 -I/home/kolev1/autotest-mfem/tux426/mfem -I/home/kolev1/autotest-mfem/tux426/mfem/../gecko/inc -o mesh/test_mesh.o
mesh/test_mesh.cpp: In function ‘void ____C_A_T_C_H____T_E_S_T____0()’:
mesh/test_mesh.cpp:19:15: error: ‘class mfem::Mesh’ has no member named ‘GetGeckoElementReordering’
          mesh.GetGeckoElementReordering(perm);
               ^
mesh/test_mesh.cpp:41:15: error: ‘class mfem::Mesh’ has no member named ‘GetGeckoElementReordering’
          mesh.GetGeckoElementReordering(perm);
               ^
mesh/test_mesh.cpp:65:22: error: ‘class mfem::Mesh’ has no member named ‘GetGeckoElementReordering’
       mesh_reordered.GetGeckoElementReordering(perm);
                      ^
make: *** [mesh/test_mesh.o] Error 1
make: *** Waiting for unfinished jobs....
Building the unit test code failed. Stop
```

tzanio commented on examples/ex1.cpp:
```diff
@@ -112,6 +112,7 @@ int main(int argc, char *argv[])
       }
    }
 
+#if 0
```
Can you please revert these and re-merge in `next`?

jakubcerveny replied:
Done, but I can't push to `next`:
```
remote: error: You're not authorized to push to this branch.
```

tzanio replied:
Ok, I pushed. Can you make sure you are in the LLNL org + you have 2FA enabled? That's required to be able to push to `next` now because of GitLab.

jakubcerveny commented:
Regarding the unit tests, this is because I renamed `GetGeckoElementReordering` -> `GetGeckoElementOrdering`. I did update the unit tests in this branch, but somehow the old tests are being run.

tzanio commented:
> Regarding the unit tests, this is because I renamed `GetGeckoElementReordering` -> `GetGeckoElementOrdering`. I did update the unit tests in this branch, but somehow the old tests are being run.

That's weird. I updated the internal repo... hopefully should be OK now.

tzanio commented:
Re-merged in `next` for testing ...

v-dobrev commented on CHANGELOG:
```diff
@@ -56,6 +56,23 @@ Miscellaneous
+-------------------------------------
+- The (p)mesh-optimizer miniapp has been updated to demonstrate mesh optimization
+  for an AMR mesh.
+
```
Some of the above text comes from different PRs and was merged here incorrectly -- e.g. there are bullets that are repeated twice.

tzanio replied:
This is fixed now

v-dobrev commented on mesh/mesh.cpp:
```diff
@@ -1345,9 +1345,9 @@ void Mesh::FinalizeQuadMesh(int generate_edges, int refine,
-void Mesh::GetGeckoElementReordering(Array<int> &ordering,
-                                     int iterations, int window,
-                                     int period, int seed)
+void Mesh::GetGeckoElementOrdering(Array<int> &ordering,
```
Do not rename things that are already in the library -- I think you should revert this change.

tzanio replied:
I generally agree, but I think the change is acceptable and desirable in this particular case

jakubcerveny replied:
Sorry for the late reply. I agree we shouldn't just change things in the API, but this case I felt was special. Since the library was never published, this code was hardly ever used by anyone outside the core team. That's why I thought the change was acceptable in this case.

tzanio commented:
Re-merged in `next` for testing ...

tzanio commented on miniapps/meshing/mesh-explorer.cpp:
```diff
@@ -221,8 +220,10 @@ int main (int argc, char *argv[])
            "x) Print sub-element stats\n"
            "f) Find physical point in reference space\n"
            "p) Generate a partitioning\n"
+           "o) Reorder elements\n"
```
Besides `e) View elements` is there a better way to visualize the ordering?

jakubcerveny replied:
I don't know of a direct way. I was just using the new "1D split" partitioning to visualize the partitions. The "curve" plot in the PR description is from Octave.
